web_src/js/render/code.js

```diff
@@ -0,0 +1,52 @@
+import {getFileViewFilePath, getFileViewFileText, createExternalLink} from '../utils/misc.js';
+import {basename, isObject} from '../utils.js';
+
+export function initFileView() {
+  if (document.querySelector('.file-view.code-view')) {
+    const fileName = basename(getFileViewFilePath());
+    if (fileName === 'package.json') {
+      processPackageJson();
+    }
+  }
+}
+
+function processPackageJson() {
+  let obj;
+  try {
+    obj = JSON.parse(getFileViewFileText());
+  } catch {
+    return;
+  }
+  if (!isObject(obj)) return;
+
+  const packages = new Set();
+
+  for (const key of [
+    'dependencies',
+    'dependenciesMeta',
+    'devDependencies',
+    'optionalDependencies',
+    'overrides',
+    'peerDependencies',
+    'peerDependenciesMeta',
+    'resolutions',
+  ]) {
+    for (const packageName of Object.keys(obj?.[key] || {})) {
+      packages.add(packageName);
+    }
+  }
+
+  // match chroma NameTag token to detect JSON object keys
+  for (const el of document.querySelectorAll('.code-inner .nt')) {
```

Comment:
It is really fragile, it doesn't seem suitable to be done on frontend.

Reply:
It's the best frontend can do, and I don't think Chroma would ever break it and even if it does, it will be graceful.

Reply:
I think it's not right to let frontend do: the JS code only sees the rendered content, and it heavily depends on Chroma behavior.

I am sure there could be a stable backend solution: fully tested and fully controllable.

Reply:
I'll personally not be doing that in backend. I'm mostly a frontend guy 😆

Reply:
> I'll personally not be doing that in backend and I think doing it in backend would massively limit future similar processings.

I do not it would limit any future processings. Instead, backend could handle all future processings better than frontend.

For example, if it needs to handle maven XML, in frontend, you need a lot of tricks to to handle the rendered XML-HTML content. But in backend, it sees the clear origin content and could add links clearly.

Reply:
Have fun implementing a VSCode-like code view in backend 😆. I think these features definitely need to be in frontend.

Reply:
General talking about "frontend" and "backend" is not clear for this case. Let's abstract the logic:

origin content (text) -> highlight (get HTML) -> split to lines (HTML-line slice) -> output to page content

So the question is: add the links between which steps?

My "backend" proposal:

* origin content (text) -> **highlight OR add links** (get HTML) -> split to lines (HTML-line slice) -> output to page content

You "frontend" proposal:

* origin content (text) -> highlight (get HTML) -> split to lines (HTML-line slice) -> output to page content -> use JS to parse the rendered HTML to add links

VSCode: VSCode has a well-designed plugin system, I haven't really looked into it, so I don't know at which step it could add links. My intuition tells me it is very unlikely to do it at the last step (eg: parse the rendered content)

Reply:
Adding only links or any other HTML content can be done in backend but my point is that if the post-processing goes beyong HTML modification, like for example a JSON "block folding" feature, this has do be done in JS and then it's better if all such post-processing code is in JS.

Reply:
Also, if we decide to render code using React or Vue, it will be much easier to migrate the JS code than to migrate it from Go.